zkevm-circuits/src/blobkzg_circuit/src/main.rs

```diff
@@ -0,0 +1,585 @@
+        let witness_values = WitnessValues::<W>::new(&input_values);
+        BlobKZGCircuit {
+            input_values,
+            witness_values,
+            _marker: std::marker::PhantomData,
+        }
+    }
+}
+
+const BIT_LEN_LIMB: usize = 68;
+const NUMBER_OF_LIMBS: usize = 4;
+
+impl<W: PrimeField, N: PrimeField> Circuit<N> for BlobKZGCircuit<W, N> {
+    type Config = BlobKZGConfig;
+    type FloorPlanner = SimpleFloorPlanner;
+
+    fn without_witnesses(&self) -> Self {
+        Self::default()
+    }
+
+    fn configure(meta: &mut ConstraintSystem<N>) -> Self::Config {
+        let integer_config = {
+            const BIT_LEN_LIMB: usize = 68;
+            const NUMBER_OF_LIMBS: usize = 4;
+            let main_gate_config = MainGate::<N>::configure(meta);
+            let range_config = {
+                let overflow_bit_lens =
+                    Rns::<W, N, NUMBER_OF_LIMBS, BIT_LEN_LIMB>::construct().overflow_lengths();
+                let composition_bit_lens = vec![BIT_LEN_LIMB / NUMBER_OF_LIMBS];
+                RangeChip::<N>::configure(
+                    meta,
+                    &main_gate_config,
+                    composition_bit_lens,
+                    overflow_bit_lens,
+                )
+            };
+            IntegerConfig::new(range_config, main_gate_config)
+        };
+
+        BlobKZGConfig { integer_config }
+    }
+
+    fn synthesize(
+        &self,
+        config: Self::Config,
+        mut layouter: impl Layouter<N>,
+    ) -> Result<(), Error> {
+        fn assign_wrong_integer<W: PrimeField, N: PrimeField>(
+            integer_chip: &IntegerChip<W, N, NUMBER_OF_LIMBS, BIT_LEN_LIMB>,
+            ctx: &mut RegionCtx<N>,
+            x: W,
+        ) -> Result<AssignedInteger<W, N, NUMBER_OF_LIMBS, BIT_LEN_LIMB>, Error> {
+            let rc_rns = std::rc::Rc::new(Rns::<W, N, NUMBER_OF_LIMBS, BIT_LEN_LIMB>::construct());
+            let integer = Integer::from_fe(x, rc_rns);
+            let value = Value::known(integer);
+            let unassigned_integer = UnassignedInteger::from(value);
+            let assigned_integer =
+                integer_chip.assign_integer(ctx, unassigned_integer, Range::Operand)?;
+            Ok(assigned_integer)
+        }
+
+        let integer_chip = {
+            let rc_rns = std::rc::Rc::new(Rns::<W, N, NUMBER_OF_LIMBS, BIT_LEN_LIMB>::construct());
+            IntegerChip::<W, N, NUMBER_OF_LIMBS, BIT_LEN_LIMB>::new(config.integer_config, rc_rns)
+        };
+        integer_chip.range_chip().load_table(&mut layouter)?;
+
+        let InputValues {
+            datapoints,
+            x,
+            w,
+            n,
+        } = self.input_values;
+        let WitnessValues {
+            evals,
+            wis,
+            xns,
+            x2is,
+            nbits,
+        } = self.witness_values;
+
+        // 0. JUT MAKE SURE X IS NOT IN THE ROOT-OF-UNITY DOMAIN (otherwise formula doesn't work)
```

Comment:
It's a matter of code style, but I usually don't perform non-constrained checks inside of a circuit. a malicious prover would just remove this code, and this adds overhead to honest provers. If we need to validate input data, we might want to do that a step before the circuit instantiation, also to help the user to debug potential errors.